### PR TITLE
Add support for PMSA003i

### PR DESCRIPTION
**Pull request by sjtrny**
## Description:

Adds support for the PMSA003i sensor, which is an i2c version of similar particulate sensors from Plantower.

Original implementation here https://github.com/sjtrny/pmsa003i

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 

https://github.com/esphome/esphome-docs/pull/973

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/pmsa003i/pmsa003i.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  sensor::Sensor *pm_2_5_sensor_{nullptr};
+  sensor::Sensor *pm_10_0_sensor_{nullptr};
+
+  sensor::Sensor *pmc_0_3_sensor_{nullptr};
+  sensor::Sensor *pmc_0_5_sensor_{nullptr};
+  sensor::Sensor *pmc_1_0_sensor_{nullptr};
+  sensor::Sensor *pmc_2_5_sensor_{nullptr};
+  sensor::Sensor *pmc_5_0_sensor_{nullptr};
+  sensor::Sensor *pmc_10_0_sensor_{nullptr};
+  sensor::Sensor *pm_size_sensor_{nullptr};
```
Comment (glmnet) [inline esphome/components/pmsa003i/pmsa003i.h:65]: Where is this used?

File: esphome/components/pmsa003i/sensor.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+CONF_PMC_0_3 = 'pmc_0_3'
+CONF_PMC_5_0 = 'pmc_5_0'
+
+CONFIG_SCHEMA = cv.Schema({
+    cv.GenerateID(): cv.declare_id(PMSA003IComponent),
+
+    cv.Optional(CONF_STANDARD_UNITS, default=True): cv.boolean,
+
+    cv.Optional(CONF_PM_1_0): sensor.sensor_schema(UNIT_MICROGRAMS_PER_CUBIC_METER,
+                                                   ICON_CHEMICAL_WEAPON, 2),
```
Comment (glmnet) [inline esphome/components/pmsa003i/sensor.py:27]: this needs rebase from dev, now sensor.sensor_schema requires 4 arguments

File: esphome/components/pmsa003i/pmsa003i.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+void PMSA003IComponent::dump_config() { LOG_I2C_DEVICE(this); }
+
+void PMSA003IComponent::update() {
+  PM25AQIData data;
+
+  bool successful_read = this->read_data_(&data);
+
+  // Update sensors
+  if (successful_read) {
+    ESP_LOGD(TAG, "Read success. Updating sensors.");
```
Comment (glmnet) [inline esphome/components/pmsa003i/pmsa003i.cpp:30]:
```suggestion
    ESP_LOGV(TAG, "Read success. Updating sensors.");
```

File: esphome/components/pmsa003i/pmsa003i.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    if (this->pmc_1_0_sensor_ != nullptr)
+      this->pmc_1_0_sensor_->publish_state(data.particles_10um);
+    if (this->pmc_2_5_sensor_ != nullptr)
+      this->pmc_2_5_sensor_->publish_state(data.particles_25um);
+    if (this->pmc_5_0_sensor_ != nullptr)
+      this->pmc_5_0_sensor_->publish_state(data.particles_50um);
+    if (this->pmc_10_0_sensor_ != nullptr)
+      this->pmc_10_0_sensor_->publish_state(data.particles_100um);
+  } else {
+    ESP_LOGD(TAG, "Read failure. Skipping update.");
```
Comment (glmnet) [inline esphome/components/pmsa003i/pmsa003i.cpp:61]: you can use https://github.com/esphome/esphome/blob/08998caabc7d592ef07bb1bdc7386ae4e697b4a0/esphome/core/component.h#L117 to set a warning here, you then need to clear it on success reads

**Review by glmnet — CHANGES_REQUESTED** [review]
Please rebase as this would fail marging on dev.

Thanks

File: esphome/components/pmsa003i/pmsa003i.h (hunk 1 above)
Comment (sjtrny) [reply esphome/components/pmsa003i/pmsa003i.h]: Nowhere. I've removed it.

File: esphome/components/pmsa003i/sensor.py (hunk 2 above)
Comment (sjtrny) [reply esphome/components/pmsa003i/sensor.py]: Done

File: esphome/components/pmsa003i/pmsa003i.cpp (hunk 3 above)
Comment (sjtrny) [reply esphome/components/pmsa003i/pmsa003i.cpp]: Done

File: esphome/components/pmsa003i/pmsa003i.cpp (hunk 4 above)
Comment (sjtrny) [reply esphome/components/pmsa003i/pmsa003i.cpp:63]: I've added this. Let me know if I made a mistake.

File: esphome/components/pmsa003i/pmsa003i.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#include "esphome/core/component.h"
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/i2c/i2c.h"
+
+namespace esphome {
+namespace pmsa003i {
+
+/**! Structure holding Plantower's standard packet **/
+// From https://github.com/adafruit/Adafruit_PM25AQI
+using PM25AQIData = struct PMSAQIdata {
```
Comment (OttoWinter) [inline esphome/components/pmsa003i/pmsa003i.h]:
```suggestion
struct PM25AQIData {
```

(typedef/using struct not needed in C++)

**Review by OttoWinter — COMMENTED** [review]
Thanks for this contribution! Code looks good to me. Just one minor thing about the `struct`, can you have a look at it

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**